src/index_map.rs

```diff
@@ -0,0 +1,432 @@
+//! [`NEIndexMap`] is a non-empty variant of [`IndexMap`].
+#![deny(clippy::correctness, clippy::suspicious)]
+#![warn(clippy::complexity, clippy::perf, clippy::style, clippy::pedantic)]
+#![allow(clippy::module_name_repetitions)]
+use std::fmt::Debug;
+use std::hash::BuildHasher;
+use std::hash::Hash;
+use std::iter;
+use std::mem;
+
+use crate::FromNonEmptyIterator;
+use crate::IntoNonEmptyIterator;
+use crate::NonEmptyIterator;
+use indexmap::Equivalent;
+use indexmap::IndexMap;
+
+/// Short-hand for constructing [`NEIndexMap`] values.
+///
+/// ```
+/// use nonempty_collections::ne_indexmap;
+///
+/// let m = ne_indexmap!{"elves" => 3000, "orcs" => 10000};
+/// assert_eq!(2, m.len());
+/// ```
+#[macro_export]
+macro_rules! ne_indexmap {
+    ($hk:expr => $hv:expr, $( $xk:expr => $xv:expr,)+) => { $crate::ne_indexmap!{$hk => $hv, $($xk => $xv),+} };
+    ($hk:expr => $hv:expr, $( $xk:expr => $xv:expr ),*) => {{
+        const CAP: usize = <[()]>::len(&[$({ stringify!($xk); }),*]);
+        let mut map = $crate::index_map::NEIndexMap::with_capacity(CAP, $hk, $hv);
+        $( map.insert($xk, $xv); )*
+        map
+    }};
+    ($hk:expr => $hv:expr) => {
+        $crate::index_map::NEIndexMap::new($hk, $hv)
+    }
+}
+
+/// A non-empty, growable [`IndexMap`].
+///
+/// ```
+/// use nonempty_collections::*;
+///
+/// let m = ne_indexmap!{"Netherlands" => 18, "Canada" => 40};
+/// assert_eq!(2, m.len());
+/// ```
+#[derive(Clone)]
+pub struct NEIndexMap<K, V, S = std::collections::hash_map::RandomState> {
+    /// The key of the ever-present element of the non-empty `IndexMap`.
+    head_key: K,
+
+    /// The value of the ever-present element of the non-empty `IndexMap`.
+    head_val: V,
+
+    /// The remaining key-value pairs, perhaps empty.
+    tail: IndexMap<K, V, S>,
+}
+
+impl<K, V, S> NEIndexMap<K, V, S> {
+    /// Returns the number of elements the map can hold without reallocating.
+    pub fn capacity(&self) -> usize {
+        self.tail.capacity() + 1
+    }
+
+    /// Returns a reference to the map's `BuildHasher`.
+    pub fn hasher(&self) -> &S {
+        self.tail.hasher()
+    }
+
+    /// An iterator visiting all elements in arbitrary order. The iterator
+    /// element type is `(&'a K, &'a V)`.
+    pub fn iter(&self) -> impl Iterator<Item = (&K, &V)> + IntoIterator<Item = (&K, &V)> {
+        // TODO implement as struct
+        iter::once((&self.head_key, &self.head_val)).chain(self.tail.iter())
+    }
+
+    // TODO iter_mut
+
+    /// An iterator visiting all keys in arbitrary order. The iterator element
+    /// type is `&'a K`.
+    ///
+    /// ```
+    /// use nonempty_collections::*;
+    ///
+    /// let m = ne_indexmap!{"Duke" => "Leto", "Doctor" => "Yueh", "Planetologist" => "Kynes"};
+    /// let v = m.keys().collect::<NEVec<_>>();
+    /// assert_eq!(nev![&"Duke", &"Doctor", &"Planetologist"], v);
+    /// ```
+    pub fn keys(&self) -> impl NonEmptyIterator<Item = &K> + IntoIterator<Item = &K> {
```

Comment:
`iter` above was marked as `Iterator` but this here is `NonEmptyIterator`. Perhaps these should be unified.